README.md
# M5_NightscoutMon

<img width="240" src="https://raw.githubusercontent.com/mlukasek/M5_NightscoutMon/master/images/M5NS_mon_2019-06-20_page1.jpg">&nbsp;&nbsp;<img width="240" src="https://raw.githubusercontent.com/mlukasek/M5_NightscoutMon/master/images/M5NS_mon_2019-06-20_page2.jpg">&nbsp;&nbsp;<img width="240" src="https://raw.githubusercontent.com/mlukasek/M5_NightscoutMon/master/images/M5NS_mon_2019-06-20_page3.jpg">

## M5Stack Nightscout monitor

##### M5Stack Nightscout monitor<br/>Copyright (C) 2018-2020 Martin Lukasek <martin@lukasek.cz>
###### This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
###### This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
###### You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>. 
###### This software uses some 3rd party libraries:<br/>IniFile by Steve Marple (GNU LGPL v2.1)<br/>ArduinoJson by Benoit BLANCHON (MIT License)<br/>IoT Icon Set by Artur Funk (GPL v3)<br/>DHT12 by Bobadas (Public domain)<br/><br/>Additions to the code:<br/>Peter Leimbach (Nightscout token)<br/>Patrick Sonnerat (Dexcom Sugarmate connection)<br/>Sulka Haro (Nightscout API queries help)<br/>Dominik Dzienia _(Refactoring, migration to Platform.IO)_

### Contents
[What is this good for?](#m5stack-nightscout-monitor-1)  
[Wiki documentation](https://github.com/mlukasek/M5_NightscoutMon/wiki)  
[Latest software revisions (CHANGELOG) **WHAT IS NEW**](CHANGELOG.md)  
[Syntax of M5NS.INI configuration file](#the-syntax-of-m5nsini-configuration-file)  
[Display](#display)  
[Buttons](#buttons)  
[Battery](#battery)  
[Installation and support](#installation-and-support)  
[Donations - support the project](#donations) :+1:


### M5Stack Nightscout Monitor

M5Stack is small development kit based on ESP32. It is in a nice plastic box, equipped with color display, micro SD card slot, 3 buttons, speaker and internal battery. It is great to monitor and display something, so I used it to monitor my daughter's glycemia. It is nice, comfortable and cheap solution.

We use Abbott Freestyle Libre, together with MiaoMiao transmitter. It transmits the glycemia values to her mobile phone (we used to use Spike on iPhone, but you can also use Tomato on iPhone or xDrip, Glimp and more on Android). The daughter’s mobile phone then sends data to the Nightscout, where we (parents) watch it using the web application or a mobile phone (Spike, Nightscout, …). We use the small M5Stack monitor with my software to monitor the glycemia during nights in our sleeping room, my wife has one in the kitchen and we take one always on our trips. It has different backlight intensity for night usage and alarms for hypo/hyper glycemia situations.

**In short it displays information from Nightscout site and provides visual and sound alarms if needed.**

Some people were asking for the software. I am not a professional programmer, I am a hardware engineer, but I work more with tables, numbers and e-mails recent years, so I was a little bit shy about the code. As there is quite a lot of people interested in something that shows current values and makes alarms during nights, I decided to release the code as an Open Source.

If you know what to do with it, just download the code, modify it any way it suits you and use it in M5Stack. If you do not know how to handle the source code, just find someone who can. It is easy, just open it in Arduino IDE, download M5Stack libraries, Arduino JSON library and build it to your M5Stack. M5Stack is in more versions, the Core is just enough. The code does not use any features of the higher models. The official M5Stack store is here https://m5stack.com/products/basic-core-iot-development-kit. The current price is about 27.95 USD (9 February 2020). You can also buy it on Aliexpress https://www.aliexpress.com/item/M5Stack-Official-Stock-Offer-ESP32-Basic-Core-Development-Kit-Extensible-Micro-Control-Wifi-BLE-IoT-Prototype/32837164440.html or elsewhere.

**You will need a microSD card.** It has to be formatted to FAT32 format and you have to put at least M5NS.INI file with configuration to the root of the microSD card. It is a good idea to put M5_NightscoutMon.jpg files to the SD card root too. You can replace M5_NightscoutMon.jpg with any 320x240 pixels picture if you want to customize your experience.

**You have to update M5NS.INI file to your requirements. As a minimum requirement you will need to update your Nightscout URL and your WiFi SSID and password.**<br/>
<br/>

### The syntax of M5NS.INI configuration file

```ini
[config]

nightscout = yoursite.herokuapp.com ; use your Nightscout site URL
token = security token              ; token for access secured Nightscout site, 
                                    ; if public Nightscout site is used, delete token line and do not use it
bootpic = /M5_NightscoutMon.jpg     ; boot picture
name = YourName                     ; display name
device_name = M5NS                  ; device name for mDNS, that can be used to  access the device from local
                                    ; intranet, use .local domain, eg. m5ns.local
disable_web_server = 0              ; To disable internal Web server enter 'disable_web_server = 1'
time_zone = 3600                    ; the time zone where you live in seconds (3600 is GMT+1, Prague)
dst = 3600                          ; your daylight saving time offset (3600s is +1 hour)
show_mgdl = 0                       ; use 0 to display values in mmol/L or 1 for mg/dl
sgv_only = 0                        ; set to 1 if you use xDrip, Spike or similar to filter out calibrations
                                    ; etc., do not use with Tomato (set 0)
default_page = 0                    ; page number displayed after startup
show_current_time = 1               ; show current time instead of last valid data time
restart_at_time = HH:MM             ; time in HH:MM format when the device will restart
restart_at_logged_errors = 0        ; restart device after particular number of errors in the log 
                                    ; (0 = do not restart)
show_COB_IOB = 1                    ; show COB and IOB, values are grayed out if COB/IOB value is 0
snooze_timeout = 30                 ; sound snooze time in minutes after press of the middle button
alarm_repeat = 5                    ; sound repeat interval in minutes (alarm repeats only if alarm/warning 
                                    ; conditions are met)
info_line = 1                       ; 0 = sensor info, 1 = button function icons, 
                                    ; 2 = loop info (Loop) + basal, 3 = OpenAPS loop info + basal
temperature_unit = 1                ; 1 = CELSIUS, 2 = KELVIN, 3 = FAHRENHEIT. Can be omitted (default Celsius).
display_rotation = 1                ; 1 = buttons down (default, can be omitted), 3 = buttons up, 
                                    ; 5 = mirror buttons up, 7 = mirror buttons down  
date_format = 0                     ; 0 = dd.mm. (default, can be omitted), 1 = MM/DD
time_format = 0                     ; 0 = 24 Hours (default, can be omitted), 1 = 12-hour AM/PM
brightness1 = 50                    ; the first (default) brightness value
brightness2 = 100                   ; the second brightness value
brightness3 = 10                    ; the third brightness value

;;;;;;;;;; following values are in mmol/L or mg/dL depending on the previous setting of "show_mgdl" key value_

yellow_low = 4.5           ; glycemia values under this value will be displayed in yellow color
yellow_high = 9            ; glycemia values over this value will be displayed in yellow color
red_low = 3.9              ; glycemia values under this value will be displayed in red color
red_high = 11              ; glycemia values over this value will be displayed in red color
snd_warning = 3.7          ; softer high beep sound will beep every 5 mins when under this value
snd_alarm = 3.0            ; strong alarm will sound every 5 min when under this value
snd_warning_high = 14.0    ; softer high beep sound will beep every 5 mins when over this value
snd_alarm_high = 20.0      ; strong alarm will sound every 5 min when over this value
snd_no_readings = 20       ; softer high beep sound will beep every 5 mins when time in minutes when last data was read 
                           ; is over this value
snd_loop_error = 1         ; 0 = no loop error alarm, 1 = stronger sound alarm to loop error
snd_warning_at_startup = 1 ; play test warning sound with its volume during startup (1 = play, 0 =  do not play)
snd_alarm_at_startup = 0   ; play test alarm sound with its volume during startup (1 = play, 0 =  do not play)
warning_volume = 20        ; volume of warning sound in range 0-100 (0=no sound, 1=silent, 100=max volume)
alarm_volume = 100         ; volume of alarm sound in range 0-100 (0=no sound, 1=silent, 100=max volume)

;;;;;;;;;; there are supported some external devices, take care and know what you are doing when you connect it

LED_strip_mode = 0         ; external RGB LED strip or internal M5Stack Fire RGB LED support
                           ; 0 = off, 1 = visualize sound, 2 = show warnings and alarms, 3 = light always
LED_strip_pin = 15         ; pin that controls LEDs, 15 = M5Stack Fire internal pin
                           ; 21 = red PORT A connector (not recommended, collides with I2C, first LEDs will blink occassionaly
                           ; 26 = black PORT B connector, 17 = blue PORT C connector
LED_strip_count = 10       ; 10 = number of M5Stack Fire internal LEDs,
                           ; 15 = 10 cm strip, 29 = 20 cm strip, 72 = 50 cm strip, 144 = 100 cm strip, 288 = 200 cm strip
LED_strip_brightness = 10  ; 1-100% brightness intesity (100 = max.) 
                           ; for more than 10 LEDs use 10%, 5% or even less otherwise loaded current will reset the M5Stack

vibration_mode = 0         ; VIBRATION MOTOR UNIT ; 0 = off, 1 = ON (vibrate during sound)
vibration_pin = 26         ; pin that controls VIBRATION UNIT motor, 26 = M5Stack Fire PORT B connector, 17 = blue PORT C connector
                           ; do not connect vibration motor to red PORT A, it will rotate always as it will be in collision with I2C
vibration_strength = 512   ; 10 bit PWM value for VIBRATION UNIT motor control, 512 = 1/2 of max. power, do not use more, full power
                           ; will reset the M5Stack, recommended range is 256-512

micro_dot_pHAT = 0         ; 0 = off, 1 = ON (display SGV and DELTA on I2C Pimoroni Micro Dot pHAT connected to I2C pins 21+22)

;;;;;;;;;;  you can connect up to 9 WiFi access points, no need to setup all, one is enough

[wlan1]
ssid = ssid1  
pass = pass1

[wlan2]  
ssid = ssid2  
pass = pass2

[wlan3]  
...
...
```

### Display

There are currently 3 display pages, that can be changed by a short press of the right button. The button does not work when the device communicated with Nightscout (small blue WiFi icon is displayed).

**Page 0 - default page**  
What you see on the display is a matter of your M5NS.INI and Nightscout configuration. By default you will see current time, date and user name from config file in upper left corner. You will see IOB (Insulin On Board) an COB (Carbs On Board) values below the name. Right from the IOB/COB value is current BG delta (difference from the last reading, trend where the BG is going). Box with time difference from last valid data from Nighscout is displayed in upper right corner. Delay of 5 mins is OK and the box has grey color, 5-15 mins is warning and the box will have white background, more than 15 mins without a valid data is error and the box background will be red. There is also possibility to set sound warning on data delay in M5NS.INI.

Glycemia (BG) value and its direction are below the basic information. There is a mini graph created from 10 last BG values right of the glycemia value. The mini graph shows only values in range 3-12 mmol/L (48-216 mg/dL) and values below/above this will be displayed on mini graph as the min/max values.

There are also icons in upper side of display. Plug or battery icon displays battery status (on newer devices only). Red clock indicated silent "snooze" mode activated. Grey or yellow (more than 5 errors) shows that there are errors in error log (last page). Blue WiFi icon in 2 sizes (2 different queries) indicated that MStack is downloading data from Nightscout.

Last line contains icons for buttons or data source or Loop information (depending on M5NS.INI key "info_line").  

**Page 1 - simple page with large BG value**  
Long distance visible page with large BG value and larger clock. Top line contains current time, icons, delta value a BG direction arrow.

Bottom info line is the same as on page 0.

**Page 2 - analog clock with temperature and humidity**  
Large analog clock with date. BG value in top left corner with time difference from the last reading. BG delta value in top right corner and arrow pointing chinge direction below it.

This screen contatins environmental values form DHT12 sensor. Requires ENV Unit or BTC Standing Base connected to MStack. The temperature is in lower left corner and humidity in lower right corner. Temperature unit can be set in M5NS.INI.

Bottom info line is the same as on page 0.

**Error Log page**
Displays last 10 record errors and total number of errors record since last restart. Occasional communication errors, such as "timeout" or "connection refused" are normal and happen sometimes due to WiFi and internet connection and Nightscout site status. If you have a lot of errors every few minutes, consider to move your M5Stack to a different place or extend WiFi signal.  
<br/>

### Buttons

The **left button** changes the backlight in the 3 steps defined in M5NS.INI file.

The **middle button** snoozes M5Stack for defined time to prevent alarm or warning sound. You can "snooze" it even before the alarm is activated. Snooze active is indicated by red clock icon in status icon area. You can press "snooze" more times to make snooze time longer. It cycles 1x - 2x - 3x - 4x - OFF and again. You have 2s to press button again, otherwise it starts from 1x. The "snooze" function is distributed to all devices in your local network with the same Nightscout user (Nightscout URL).

Short press of the **right button** switches display pages. Long press (4 seconds) puts M5Stack to the sleep as it is a little bit tricky to double click standard red power button.  
<br/>

### Battery

Battery icon feature works only on newer M5Stack units. Here are my battery measurements and battery icon statuses:

| internal battery total                   |  43 min |
| ---------------------------------------- | ------: |
| plug icon _(full charge or plugged in)_  |   1 min |
| gray full battery icon                   |  13 min |
| white 2/3 battery icon                   |  20 min |
| yellow 1/3 battery icon                  |   8 min |
| red empty battery icon                   |   1 min |

---

| extended battery total                   |  7h 49 min |
| ---------------------------------------- | ---------: |
| plug icon _(full charge or plugged in)_  |        53m |
| gray full battery icon                   | 2 h 59 min |
| white 2/3 battery icon                   | 1 h 55 min |
| yellow 1/3 battery icon                  | 1 h 48 min |
| red empty battery icon                   |     14 min |

### Installation and support

Installation can be done by download or git clone the code to the Arduino IDE environment. You have to download ESP32 board and M5Stack libraries, Arduino JSON library and build it to you M5Stack.

Another easier possibility is to download latest [M5Burner release](https://github.com/mlukasek/M5_NightscoutMon/releases). It is Windows executable with binary M5Stack firmware included. Just unzip it, start the M5Burner.exe, choose COM port where your M5Stack is connected and burn the firmware. Simple. Remember to add you microSD card with M5NS.INI configuration file. See the [wiki](https://github.com/mlukasek/M5_NightscoutMon/wiki) for details.

There is a Facebook group [M5STACK NIGHTSCOUT](https://www.facebook.com/groups/606295776549008/) where you can get support and installation guides in several languages prepared by members of the M5Stack Nightscout community. Please check the [Files section](https://www.facebook.com/groups/606295776549008/files/) first and search the group before asking questions. A lot of questions have been answered already. Big thanks to Patrick Sonnerat, Didier Frétigné, Peter Leimbach and more...

### VSCode & Platform.IO Development

* You can use VSCode & Platform.IO besides Arduino IDE for development.
* For development use free [Visual Studio Code](https://code.visualstudio.com/)
* Install Platform.IO addon in VSCode:
  * open `Extensions` from left sidebar
  * search for `Platform.IO` in extension search bar
  * click green `Install` button and wait till it is installed
* Unpack the M5_NightscoutMon_PlatformIO.zip to PlatformIO projects folder.
* Copy the latest sources from GitHub root to PlatformIO project src folder (overwrite) and rename M5_NightscoutMon.ino to M5_NightscoutMon.cpp (the .zip dows not have to be updated to the latest version).
* You can open project directly from VSCode `File -> Open folder...` menu, or from `PIO Home` with `Open project` button.
* Project is already configured for Platform.IO & VSCode. During build it will download and configure all dependencies.
* To **build** project press `PlatformIO: Build` button (it is white check mark icon `âś“` on the blue bottom toolbar)
* To **upload** your project to the m5stack board:
  * Connect your board via USB-C cable to your Laptop/PC
  * press `PlatformIO: Upload` button (it is white right arrow icon `â†’` on the blue bottom toolbar)

### Donations

If you find my project useful, I will appreciate if you donate me few bucks for further development or buy me a glass of beer or wine :wine_glass:.

https://paypal.me/8bity

I also collect and repair vintage [8-bit computers](http://www.8bity.cz/), so if you have some really old computer, I could give it a new home and care :smile:. Thank you!


CHANGELOG.md
# M5Stack Nightscout monitor project Changelog

## Revisions


### 2 October 2022

* Updates for Railway (longer token, accept "trend" in addition to "direction", response count limited to 10).  

### 28 May 2021

* Previous delta (plugins) value display corrected. Nightscount read postponed from 300 to 305 seconds after the last update to allow plugins (delta, iob, cob, ...) to update.  
* Corrected error when "/" character was at the end of Nightscout URL.  
* Redirection for HTTP codes 301 and 302.  
* HTTPClient and WiFiClientSecure returned back to readNightscout() function.  

### 21 April 2021

* setReuse (keep-alive) returned back (removed) as it was causing crashes in HTTPClient library for Sugarmate.

### 20 April 2021

* Stability and server load improvements.  
* Binary built with older versions of M5Stack board 1.0.6 and M5Stack library 0.3.0 as newer versions did not recover after HTTP error.  
* Reusing of http connection disabled (keep-alive).  
* HTTPS GET uses root certificate “DigiCert High Assurance EV Root CA” for herokuapp.com site. It behaves „the old way“ for other servers.  
* Reduce Nightscout connections (NS is queried for update only once every minute and only if time from the last reading is bigger than 5 minutes).  
* Free memory and Up time added to the info page for diagnostic purposes.  
* Date format update (added 0 to days bellow 10).  

### 12 March 2021

* SoftAP mode added for configuration possibility on a new WiFi (hold button A and then click reset on M5Stack models with mechanical buttons, touch button A while you see CONFIG progress during boot on M5Stack Core2). 
* No SD card needed. When you save configuration from the web interface, it is saved to both - SD card and internal flash memory. You can then remove the SD card and reboot. To archive your settings to SD card, insert SD card and save configuration from web interface.  
* If you start blank M5Stack just programmed with M5 Nightscout Monitor firmware, it will enter the SoftAP mode directly for easy web configuration.  
* SSID name length extended to 63 characters.  

### 31 January 2021

* Added SHT30 compatibility, so now temperature and humidity works with newer version of BTC TICKER stand and ENV.II Unit.  

### 6 December 2020

* Mini graph corrected for Sugarmate users.  
* Alarm sound to loop error is now possible.  
* Support for 12-hour AM/PM time format.  
* Vibration motor unit support for hearing impaired.  
* Flexible RGB LED strip or M5Stack Fire internal RGB LEDs support.  
* Micro Dot pHAT support.  
* Core2 online update possibility (on next update, now you have to use M5Burner or build from sources).  

### 24 July 2020

* Delta displayed in correct M5NS.INI units regardles from Nightscout settings. (romkuru)  
* Corrected "bellow->below" typo in web interface. (guydavies)  
* Replacing Unicode TAB character "/u000b" defined by Medtronic by space to correctly parse JSON without enabling Unicode.  
* Added SHT30 compatibility, so now temperature and humidity works with newer version of BTC TICKER stand and ENV.II Unit.  

### 12 May 2020

* Delta displayed in correct M5NS.INI units regardles from Nightscout settings. (romkuru)  
* Corrected "bellow->below" typo in web interface. (guydavies)  

### 12 April 2020

* WiFi passwords are now hidden on all web pages and forms.  
* "snooze" part completely reworked.  
* Possibility to multiply alarm/warning sound snooze time by multiple presses of middle button. The button has to be pressed in less than 2s after the last press. The steps are 1x-2x-3x-4x-OFF and again.  
* Snooze (middle button) from one device works on all devices on the same network subnet with the same user (Nightscout URL). UDP broadcast is distrubuted in local network subnet.  

### 13 February 2020

* Improved time handling when Epoch in milliseconds has decimal places (CGMBLEKit).
* Added version number to the latest (log) page.  
* Some minor diagnostics monitor/log updates.  

### 09 February 2020

* Support for OpenAPS loop (set info_line = 3 in M5NS.INI).  
* Corrected error in token implementation for IOB, COB, Loop, OpenAPS, basal.  
* Corrected an error in shifting SGV history graph buffer.  
* Commented out some unused variables  
* Renamed WiFiMulti to WiFiMultiple. (Dominik Dzienia)  
* Some externals moved to extern.h  (Dominik Dzienia)  
* Sources are now possible to compile in VSCode & Platform.IO, which is about 4 times faster, has better syntax highlighting, ... Just rename .ino to .cpp (Dominik Dzienia)  
* Cleanup of README.md, changelog refactored to CHANGELOG.md (Dominik Dzienia)  

### 26 December 2019

* HTTP Server now supports full configuration edit and save.  
* WebConfig sources moved to separate files.  
* Backup copy of `M5NS.INI` is copied to M5NS.BAK during save.  

### 20 October 2019

This is mostly test release with new "geek" features. Should be mostly for people with specific troubles and early adopters.

* Maximum password length extended to 63 characters.
* Possibility to connect to open WiFi network (without password).  
device_name key added to `M5NS.INI` (default is M5NS).  
* mDNS added, you can connect to M5 Nightscout monitor by device_name.local (default M5NS.local).  
* mDNS name and IP address is displayed on Error log page.  
* Experimental web server added. Display configuration and check the current and the latest firmware version.  
* Online update option from internal web page.  

> Please note, there is no security implemented yet. Internal web page is on unsecured port 80. Updates are going from server port 80, no SLL used currently. Use on your own risk.  

### 10 October 2019


microdot.cpp
#include "microdot.h"

const byte col[480] = {                     // even digits (2,4,6) have column bit map for each column of 5x7 matrix
    0x00, 0x00, 0x00, 0x00, 0x00, // (space)  00
    0x00, 0x00, 0x5F, 0x00, 0x00, // !        01
    0x00, 0x07, 0x00, 0x07, 0x00, // "        02
    0x14, 0x7F, 0x14, 0x7F, 0x14, // #        03
    0x24, 0x2A, 0x7F, 0x2A, 0x12, // $        04
    0x23, 0x13, 0x08, 0x64, 0x62, // %        05
    0x36, 0x49, 0x55, 0x22, 0x50, // &        06
    0x00, 0x05, 0x03, 0x00, 0x00, // '        07
    0x00, 0x1C, 0x22, 0x41, 0x00, // (        08
    0x00, 0x41, 0x22, 0x1C, 0x00, // )        09
    0x08, 0x2A, 0x1C, 0x2A, 0x08, // *        0a
    0x08, 0x08, 0x3E, 0x08, 0x08, // +        0b
    0x00, 0x50, 0x30, 0x00, 0x00, // ,        0c
    0x08, 0x08, 0x08, 0x08, 0x08, // -        0d
    0x00, 0x60, 0x60, 0x00, 0x00, // .        0e
    0x20, 0x10, 0x08, 0x04, 0x02, // /        0f
    0x3E, 0x51, 0x49, 0x45, 0x3E, // 0        10
    0x00, 0x42, 0x7F, 0x40, 0x00, // 1        11
    0x42, 0x61, 0x51, 0x49, 0x46, // 2        12
    0x21, 0x41, 0x45, 0x4B, 0x31, // 3        13
    0x18, 0x14, 0x12, 0x7F, 0x10, // 4        14
    0x27, 0x45, 0x45, 0x45, 0x39, // 5        15
    0x3C, 0x4A, 0x49, 0x49, 0x30, // 6        16
    0x01, 0x71, 0x09, 0x05, 0x03, // 7        17
    0x36, 0x49, 0x49, 0x49, 0x36, // 8        18
    0x06, 0x49, 0x49, 0x29, 0x1E, // 9        19
    0x00, 0x36, 0x36, 0x00, 0x00, // :        1a
    0x00, 0x56, 0x36, 0x00, 0x00, // ;        1b
    0x00, 0x08, 0x14, 0x22, 0x41, // <        1c
    0x14, 0x14, 0x14, 0x14, 0x14, // =        1d
    0x41, 0x22, 0x14, 0x08, 0x00, // >        1e
    0x02, 0x01, 0x51, 0x09, 0x06, // ?        1f
    0x32, 0x49, 0x79, 0x41, 0x3E, // @        20
    0x7E, 0x11, 0x11, 0x11, 0x7E, // A        21
    0x7F, 0x49, 0x49, 0x49, 0x36, // B        22
    0x3E, 0x41, 0x41, 0x41, 0x22, // C        23
    0x7F, 0x41, 0x41, 0x22, 0x1C, // D        24
    0x7F, 0x49, 0x49, 0x49, 0x41, // E        25
    0x7F, 0x09, 0x09, 0x01, 0x01, // F        26
    0x3E, 0x41, 0x41, 0x51, 0x32, // G        27
    0x7F, 0x08, 0x08, 0x08, 0x7F, // H        28
    0x00, 0x41, 0x7F, 0x41, 0x00, // I        29
    0x20, 0x40, 0x41, 0x3F, 0x01, // J        2a
    0x7F, 0x08, 0x14, 0x22, 0x41, // K        2b
    0x7F, 0x40, 0x40, 0x40, 0x40, // L        2c
    0x7F, 0x02, 0x04, 0x02, 0x7F, // M        2d
    0x7F, 0x04, 0x08, 0x10, 0x7F, // N        2e
    0x3E, 0x41, 0x41, 0x41, 0x3E, // O        2f
    0x7F, 0x09, 0x09, 0x09, 0x06, // P        30
    0x3E, 0x41, 0x51, 0x21, 0x5E, // Q        31
    0x7F, 0x09, 0x19, 0x29, 0x46, // R        32
    0x46, 0x49, 0x49, 0x49, 0x31, // S        33
    0x01, 0x01, 0x7F, 0x01, 0x01, // T        34
    0x3F, 0x40, 0x40, 0x40, 0x3F, // U        35
    0x1F, 0x20, 0x40, 0x20, 0x1F, // V        36
    0x7F, 0x20, 0x18, 0x20, 0x7F, // W        37
    0x63, 0x14, 0x08, 0x14, 0x63, // X        38
    0x03, 0x04, 0x78, 0x04, 0x03, // Y        39
    0x61, 0x51, 0x49, 0x45, 0x43, // Z        3A
    0x00, 0x00, 0x7F, 0x41, 0x41, // [        3B
    0x02, 0x04, 0x08, 0x10, 0x20, // "\"      3C
    0x41, 0x41, 0x7F, 0x00, 0x00, // ]        3D
    0x04, 0x02, 0x01, 0x02, 0x04, // ^        3E
    0x40, 0x40, 0x40, 0x40, 0x40, // _        3F
    0x00, 0x01, 0x02, 0x04, 0x00, // `        40
    0x20, 0x54, 0x54, 0x54, 0x78, // a        41
    0x7F, 0x48, 0x44, 0x44, 0x38, // b        42
    0x38, 0x44, 0x44, 0x44, 0x20, // c        43
    0x38, 0x44, 0x44, 0x48, 0x7F, // d        44
    0x38, 0x54, 0x54, 0x54, 0x18, // e        45
    0x08, 0x7E, 0x09, 0x01, 0x02, // f        46
    0x08, 0x14, 0x54, 0x54, 0x3C, // g        47
    0x7F, 0x08, 0x04, 0x04, 0x78, // h        48
    0x00, 0x44, 0x7D, 0x40, 0x00, // i        49
    0x20, 0x40, 0x44, 0x3D, 0x00, // j        4A
    0x00, 0x7F, 0x10, 0x28, 0x44, // k        4B
    0x00, 0x41, 0x7F, 0x40, 0x00, // l        4C
    0x7C, 0x04, 0x18, 0x04, 0x78, // m        4D
    0x7C, 0x08, 0x04, 0x04, 0x78, // n        4E
    0x38, 0x44, 0x44, 0x44, 0x38, // o        4F
    0x7C, 0x14, 0x14, 0x14, 0x08, // p        50
    0x08, 0x14, 0x14, 0x18, 0x7C, // q        51
    0x7C, 0x08, 0x04, 0x04, 0x08, // r        52
    0x48, 0x54, 0x54, 0x54, 0x20, // s        53
    0x04, 0x3F, 0x44, 0x40, 0x20, // t        54
    0x3C, 0x40, 0x40, 0x20, 0x7C, // u        55
    0x1C, 0x20, 0x40, 0x20, 0x1C, // v        56
    0x3C, 0x40, 0x30, 0x40, 0x3C, // w        57
    0x44, 0x28, 0x10, 0x28, 0x44, // x        58
    0x0C, 0x50, 0x50, 0x50, 0x3C, // y        59
    0x44, 0x64, 0x54, 0x4C, 0x44, // z        5A
    0x00, 0x08, 0x36, 0x41, 0x00, // {        5B
    0x00, 0x00, 0x7F, 0x00, 0x00, // |        5C
    0x00, 0x41, 0x36, 0x08, 0x00, // }        5D
    0x08, 0x08, 0x2A, 0x1C, 0x08, // ->       5E
    0x08, 0x1C, 0x2A, 0x08, 0x08}; // <-      5F

MicroDot::MicroDot(uint8_t address) {
  aDig56 = address;
  aDig34 = address+1;
  aDig12 = address+2;
}

void MicroDot::begin() {
  Wire.begin();

  for(int i=0; i<6; i++)
    buf[i]=32;
    
  delay(15);

  Wire.beginTransmission(aDig12);
  Wire.write(0xFF);                 // reset config
  Wire.write(0x00);                 
  Wire.endTransmission(); 
  
  Wire.beginTransmission(aDig12);
  Wire.write(0x00);                 // write to configuration register
  Wire.write(0x18);                 // 8x8
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig34);
  Wire.write(0xFF);                 // reset config
  Wire.write(0x00);                 
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig34);
  Wire.write(0x00);                 // write to configuration register
  Wire.write(0x18);                 // 8x8
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig56);
  Wire.write(0xFF);                 // reset config
  Wire.write(0x00);                 
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig56);
  Wire.write(0x00);                 // write to configuration register
  Wire.write(0x18);                 // 8x8
  Wire.endTransmission(); 

  setBrightness(0x40);
  
  delay(15);
}

void MicroDot::setBrightness(uint8_t brightness) {
  uint8_t current = 0x00; // 40 mA
  
  Wire.beginTransmission(aDig12);
  Wire.write(0x0D);                 // write to Lighting  Effect  Register
  Wire.write(current);              // write current
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig12);
  Wire.write(0x19);                 // write to PWM Register 
  Wire.write(brightness);           // write row data
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig34);
  Wire.write(0x0D);                 // write to Lighting  Effect  Register
  Wire.write(current);              // write current
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig34);
  Wire.write(0x19);                 // write to PWM Register 
  Wire.write(brightness);           // write row data
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig56);
  Wire.write(0x0D);                 // write to Lighting  Effect  Register
  Wire.write(current);              // write current
  Wire.endTransmission(); 

  Wire.beginTransmission(aDig56);
  Wire.write(0x19);                 // write to PWM Register 
  Wire.write(brightness);           // write row data
  Wire.endTransmission(); 
}

void MicroDot::update(uint8_t address) {
  Wire.beginTransmission(address);
  Wire.write(0x0c);
  Wire.write(0xff);
  Wire.endTransmission();
}

void MicroDot::writeDigit(uint8_t digit, uint8_t ch) {
  byte row[7];
  uint8_t x = (ch & 0x7F)-32;
  if(digit>6)
    return;
  buf[digit-1] = ch;
  switch(digit) {
    case 1:
      Wire.beginTransmission(aDig12);
      Wire.write(0x0E);
      for(int coly=0; coly<5; coly++)
        Wire.write(col[x*5 + coly]);
      Wire.write(0x00);
      Wire.write(0x00);
      if((ch & 0x80)!=0)
        Wire.write(0x40);
      else
        Wire.write(0x00);
      Wire.endTransmission();
      update(aDig12);
      break;

    case 2:
      col2RowConv(x, row);
      Wire.beginTransmission(aDig12);
      Wire.write(0x01);
      for(int rowx=0; rowx<7; rowx++) {
        if(((ch & 0x80)!=0) && rowx==6)
          Wire.write(row[rowx] | 0x80);
        else
          Wire.write(row[rowx]);
      }
      Wire.endTransmission();
      update(aDig12);
      break;

    case 3:
      Wire.beginTransmission(aDig34);
      Wire.write(0x0E);
       for (int coly=0; coly<5; coly++)
         Wire.write (col[x*5 + coly]);
      Wire.write(0x00);
      Wire.write(0x00);
      if((ch & 0x80)!=0)
        Wire.write(0x40);
      else
        Wire.write(0x00);
      Wire.endTransmission();
      update(aDig34);
      break;
    
    case 4:
      col2RowConv(x, row);
      Wire.beginTransmission(aDig34);
      Wire.write(0x01);
      for(int rowx=0; rowx<7; rowx++) {
        if(((ch & 0x80)!=0) && rowx==6)
          Wire.write(row[rowx] | 0x80);
        else
          Wire.write(row[rowx]);
      }
      Wire.endTransmission();
      update(aDig34);
      break;

    case 5:
      Wire.beginTransmission(aDig56);
      Wire.write(0x0E);
      for(int coly=0; coly<5; coly++)
        Wire.write(col[x*5 + coly ]);
      Wire.write(0x00);
      Wire.write(0x00);
      if((ch & 0x80)!=0)
        Wire.write(0x40);
      else
        Wire.write(0x00);
      Wire.endTransmission();
      update(aDig56);
      break;

    case 6:
      col2RowConv(x, row);
      Wire.beginTransmission(aDig56);
      Wire.write(0x01);
      for(int rowx=0; rowx<7; rowx++) {
        if(((ch & 0x80)!=0) && rowx==6)
          Wire.write(row[rowx] | 0x80);
        else
          Wire.write(row[rowx]);
      }
      Wire.endTransmission();
      update(aDig56);
  }
}

void MicroDot::col2RowConv(byte digit, uint8_t* row) {
  byte rowindex = 0;
  for(rowindex=0; rowindex < 7; rowindex++) {
    row[rowindex] = 0;
    row[rowindex] =  (bitRead(col[digit * 5 + 0], rowindex) * 1)
                    +(bitRead(col[digit * 5 + 1], rowindex) * 2)
                    +(bitRead(col[digit * 5 + 2], rowindex) * 4)
                    +(bitRead(col[digit * 5 + 3], rowindex) * 8)
                    +(bitRead(col[digit * 5 + 4], rowindex) * 16);
  }
}

void MicroDot::clear() {
  for(int i=0; i<6; i++)
    writeDigit(i+1, 32);
}

void MicroDot::writeString(const char* buf) {
  for(int i=0; i<6; i++) {
    if(i<strlen(buf))
      writeDigit(i+1, buf[i]);
    else      
      writeDigit(i+1, 32);
  }
}
